dspace-spring-rest/src/main/java/org/dspace/app/rest/DiscoverableEndpointsService.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.springframework.hateoas.Link;
+import org.springframework.stereotype.Service;
+
+/**
+ * A Service able to list all the discoverable endpoints in our REST
+ * application. Endpoints need to register their managed endpoints. The service
+ * is responsible to check conflict and priorities
+ *
+ * @author Andrea Bollini (andrea.bollini at 4science.it)
+ * @author Tim Donohue
+ */
+@Service
+public class DiscoverableEndpointsService {
+	/**
+	 * Contains all the registeredEndpoints as received by the single controller
+	 */
+	private Map<Object, List<Link>> registeredEndpoints = new HashMap<Object, List<Link>>();
+
+	/**
+	 * Is the computed list of discoverableEndpoints valid?
+	 */
+	private boolean initialized = false;
+
+	private List<Link> discoverableEndpoints = new ArrayList<Link>();
+
+	public void register(Object controller, List<Link> links) {
+		synchronized (this) {
+			initialized = false;
```

Comment:
no, it should be done after the creation of the result lists in the getDiscoverableEndpoints method (see comment below)